### PR TITLE
POST request to start session with GovPay

### PR DESCRIPTION
**Pull request by jphillips2121**
Create POST request to start session with GovPay and return JourneyURL to access session. 

Resolves: CPS-143

* [X] New feature

### Pull request checklist

* [X] I have added unit tests for new code that I have added
* [X] I have added/updated functional tests where appropriate
* [X] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/go.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/golang/go/wiki/CodeReviewComments)__

### CODE REVIEW

File: service/external_paysession.go [hunk 1]
```diff
@@ -4,43 +4,80 @@ import (
...
 
-	paymentJourney := &models.ExternalPaymentJourney{}
-	//TODO: Return next_url from GovPay, hardcoded at the moment
-	paymentJourney.NextURL = "http://gov.uk/paymentjourney"
-	w.Header().Set("Content-Type", "application/json")
-	w.WriteHeader(http.StatusOK)
-	err = json.NewEncoder(w).Encode(paymentJourney)
+	log.ErrorR(req, fmt.Errorf("payment method, [%s], for resource [%s] not recognised", paymentSession.PaymentMethod, id))
+	w.WriteHeader(http.StatusBadRequest)
+	return
```
Comment (emead) [inline service/external_paysession.go]: Delete this return

File: service/external_paysession.go [hunk 2]
```diff
@@ -4,43 +4,80 @@ import (
...
+		return 0, fmt.Errorf("amount [%s] format incorrect", decimalPayment)
+	}
+
+	if strings.Contains(decimalPayment, ".") {
+		c := strings.Replace(decimalPayment, ".", "", -1)
+		pencePayment, _ := strconv.ParseInt(c, 10, 64)
+		return int(pencePayment), nil
+	}
+
+	pencePayment, _ := strconv.ParseInt(decimalPayment, 10, 64)
```
Comment (emead) [inline service/external_paysession.go]: Error not handled

File: service/external_paysession.go [hunk 3]
```diff
@@ -4,43 +4,80 @@ import (
...
 	}
+
+	matched := r.MatchString(decimalPayment)
+	if !matched {
+		return 0, fmt.Errorf("amount [%s] format incorrect", decimalPayment)
+	}
+
+	if strings.Contains(decimalPayment, ".") {
+		c := strings.Replace(decimalPayment, ".", "", -1)
+		pencePayment, _ := strconv.ParseInt(c, 10, 64)
```
Comment (emead) [inline service/external_paysession.go]: Error not handled

File: service/gov_pay_test.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
+package service
+
+import (
+	"fmt"
+	"testing"
+
+	"gopkg.in/jarcoal/httpmock.v1"
+
+	"github.com/companieshouse/payments.api.ch.gov.uk/models"
```
Comment (emead) [inline service/gov_pay_test.go]: Please group all the external imports together

File: service/external_paysession.go [hunk 5]
```diff
@@ -4,43 +4,80 @@ import (
...
-	w.Header().Set("Content-Type", "application/json")
-	w.WriteHeader(http.StatusOK)
-	err = json.NewEncoder(w).Encode(paymentJourney)
+	log.ErrorR(req, fmt.Errorf("payment method, [%s], for resource [%s] not recognised", paymentSession.PaymentMethod, id))
+	w.WriteHeader(http.StatusBadRequest)
+	return
+
+}
+
+func convertToPenceFromDecimal(decimalPayment string) (int, error) {
```
Comment (emead) [inline service/external_paysession.go]: As discussed:
* refactor `getTotalAmount` to return totalAmount.StringFixed(2)
* remove the regex checking from this function
* refactor the conversion to pence

File: service/external_paysession_test.go [hunk 6]
```diff
@@ -1,52 +1,201 @@
...
-	"io/ioutil"
+	"encoding/json"
+	"fmt"
 	"net/http"
 	"net/http/httptest"
 	"testing"
 
+	"gopkg.in/jarcoal/httpmock.v1"
+
+	"github.com/companieshouse/payments.api.ch.gov.uk/models"
```
Comment (emead) [inline service/external_paysession_test.go]: Please group the external imports together

File: service/external_paysession.go [hunk 7]
```diff
@@ -4,43 +4,80 @@ import (
...
+		log.ErrorR(req, err)
 		return
 	}
 
-	if incomingExternalPaymentJourneyRequest.PaymentMethod != "GovPay" {
-		log.ErrorR(req, fmt.Errorf("payment method not recognised: %v", incomingExternalPaymentJourneyRequest.PaymentMethod))
-		w.WriteHeader(http.StatusBadRequest)
+	paymentJourney := &models.ExternalPaymentJourney{}
+
+	if paymentSession.PaymentMethod == "GovPay" {
```
Comment (emead) [inline service/external_paysession.go]: I know there's only one condition at the moment, but a `switch` statement would be better for adding more payment methods in the future

File: service/gov_pay.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+func returnNextURLGovPay(paymentResourceData *models.PaymentResourceData, id string, cfg *config.Config) (string, error) {
+	var govPayRequest models.OutgoingGovPayRequest
+
+	amountToPay, err := convertToPenceFromDecimal(paymentResourceData.Amount)
+	if err != nil {
+		return "", fmt.Errorf("error converting amount to pay to pence: [%s]", err)
+	}
+
+	govPayRequest.Amount = amountToPay
+	govPayRequest.Description = "Companies House Payment" //Hardcoded for now as a payment description isn't saved anywhere
```
Comment (emead) [inline service/gov_pay.go]: Shoud this be a `TODO` or `FIXME`?

File: service/gov_pay.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	govPayRequest.Reference = paymentResourceData.Reference
+	govPayRequest.ReturnURL = cfg.PaymentsWebURL + "/payments/" + id + "/paymentStatus"
```
Comment (emead) [inline service/gov_pay.go]: Should this full URL be provided by payments.web?

File: models/gov_pay.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+type GovPayLinks struct {
+	Self        Self        `json:"self"`
+	NextURL     NextURL     `json:"next_url"`
+	NextURLPost NextURLPost `json:"next_url_post"`
+	Events      Events      `json:"events"`
+	Refunds     Refunds     `json:"refunds"`
+	Cancel      Cancel      `json:"cancel"`
+}
+
+type Self struct {
```
[the remark this answers is not in the document]
Comment (jphillips2121) [reply models/gov_pay.go]: Added

File: models/gov_pay.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+	HREF   string `json:"href"`
+	Method string `json:"method"`
+}
+
+type NextURL struct {
```
[the remark this answers is not in the document]
Comment (jphillips2121) [reply models/gov_pay.go]: Added

File: models/gov_pay.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	HREF   string `json:"href"`
+	Method string `json:"method"`
+}
+
+type NextURLPost struct {
```
[the remark this answers is not in the document]
Comment (jphillips2121) [reply models/gov_pay.go]: Added

File: models/gov_pay.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	PostType string `json:"type"`
+	Params   Params `json:"params"`
+	HREF     string `json:"href"`
+	Method   string `json:"method"`
+}
+
+type Params struct {
```
[the remark this answers is not in the document]
Comment (jphillips2121) [reply models/gov_pay.go]: Added

File: models/gov_pay.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+	ChargeTokenId string `json:"chargeTokenId"`
```
[the remark this answers is not in the document]
Comment (jphillips2121) [reply models/gov_pay.go]: Added

File: models/gov_pay.go [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+type Events struct {
```
[the remark this answers is not in the document]
Comment (jphillips2121) [reply models/gov_pay.go]: Added

File: models/gov_pay.go [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+	HREF   string `json:"href"`
+	Method string `json:"method"`
+}
+
+type Refunds struct {
```
[the remark this answers is not in the document]
Comment (jphillips2121) [reply models/gov_pay.go]: Added

File: models/gov_pay.go [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+	HREF   string `json:"href"`
+	Method string `json:"method"`
+}
+
+type Cancel struct {
```
[the remark this answers is not in the document]
Comment (jphillips2121) [reply models/gov_pay.go]: Added

File: models/gov_pay.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (emead) [reply models/gov_pay.go]: Not updated.

File: service/external_paysession.go [hunk 18]
```diff
@@ -4,43 +4,66 @@ import (
...
 	}
+
+	log.ErrorR(req, fmt.Errorf("payment method, [%s], for resource [%s] not recognised", paymentSession.PaymentMethod, id))
+	w.WriteHeader(http.StatusBadRequest)
+}
+
+func convertToPenceFromDecimal(decimalPayment string) (int, error) {
+	c := strings.Replace(decimalPayment, ".", "", -1)
+	pencePayment, _ := strconv.ParseInt(c, 10, 64)
+	return int(pencePayment), nil
```
Comment (emead) [inline service/external_paysession.go]: Instead of these last 2 lines, try:
`strconv.Atoi(pencePayment)`

File: service/external_paysession.go [hunk 19]
```diff
@@ -4,43 +4,66 @@ import (
...
 		return
 	}
+
+	log.ErrorR(req, fmt.Errorf("payment method, [%s], for resource [%s] not recognised", paymentSession.PaymentMethod, id))
+	w.WriteHeader(http.StatusBadRequest)
+}
+
+func convertToPenceFromDecimal(decimalPayment string) (int, error) {
+	c := strings.Replace(decimalPayment, ".", "", -1)
+	pencePayment, _ := strconv.ParseInt(c, 10, 64)
```
Comment (emead) [inline service/external_paysession.go]: All errors should be handled

File: service/external_paysession.go [hunk 20]
```diff
@@ -56,14 +56,15 @@ func (service *PaymentService) CreateExternalPaymentJourney(w http.ResponseWrite
...
+
+		w.WriteHeader(http.StatusBadRequest)
+	}
 }
 
 func convertToPenceFromDecimal(decimalPayment string) (int, error) {
-	c := strings.Replace(decimalPayment, ".", "", -1)
-	pencePayment, _ := strconv.ParseInt(c, 10, 64)
-	return int(pencePayment), nil
+	pencePayment := strings.Replace(decimalPayment, ".", "", -1)
```
Comment (emead) [inline service/external_paysession.go]: Change the count from `-1` to `1`
(There can only be one decimal point)

**Pull request merged**